app/server.js

```diff
@@ -8,6 +8,222 @@ var uid = require('rand-token').uid;
 
 const PORT = 3001;
 
-http.createServer(function (request, response) {
+//api key storage
+const VALID_API_KEYS = ["98738203-8dsj-0983-9f72-di29dk38djk3", "8di3k99l-5305-8dk3-1849-zjdf938jfj98"];
+const TOKEN_VALIDITY_TIMEOUT = 15 * 60 * 10000;
 
-}).listen(PORT);
\ No newline at end of file
+//state-holding variables, cart is kept inside currentUser via /POST /api/me/cart
+let brands = [];
+let products = [];
+let users = [];
+let currentUser = {};
+let accessTokens = [];
+let failedLoginAttempts = {};
+
+//Setup for the router
+let myRouter = Router();
+myRouter.use(bodyParser.json());
+
+//Setup for server
+let server = http.createServer(function (request, response) {
+  if(!VALID_API_KEYS.includes(request.headers["x-authentication"])) {
+    response.writeHead(401, "You need to have a valid API key to use this API")
+  }
+  response.writeHead(200);
+    myRouter(request, response, finalHandler(request, response))
+});
+
+server.listen(PORT, err => {
+  if (err) throw err;
+  console.log(`server running on port port ${PORT}`);
+  //populate brands
+  brands = JSON.parse(fs.readFileSync("initial-data/brands.json", "utf-8"));
+  //populate products
+  products = JSON.parse(fs.readFileSync("initial-data/products.json", "utf-8"));
+  //populate users
+  users = JSON.parse(fs.readFileSync("initial-data/users.json", "utf-8"));
+})
+//GET all brands
+myRouter.get("/api/brands", function(request,response) {
+  if(!request) {
+    return console.log('request not defined')
+  }
+  let brandsToReturn = [];
+  brandsToReturn = brands;
+  response.writeHead(200, { "Content-Type": "application/json" });
+  return response.end(JSON.stringify(brandsToReturn));
+})
+
+//GET all products of a single brand (by categoryId)
+myRouter.get("/api/brands/:id/products", (request, response) =>{
+  const { id } = request.params;
+  const category = brands.find(category => category.id == id);
+  if (!category) {
+    response.writeHead(404, "That brand does not exist");
+    return response.end()
+  }
+  response.writeHead(200, { "Content-Type": "application/json" });
+  const relatedProducts = products.filter(
+    products => products.categoryId === id
+  )
+  return response.end(JSON.stringify(relatedProducts));
+})
+//GET all products
+myRouter.get("/api/products", function(request,response) {
+  if(!request) {
+    return console.log('request not defined')
+  }
+  let productsToReturn = [];
+  productsToReturn = products;
+  // console.log('products to return', productsToReturn);
+  response.writeHead(200, { "Content-Type": "application/json" });
+  return response.end(JSON.stringify(productsToReturn));
+})
+
+//Functions to help manage failed requests by username
+let getNumFailedLoginRequestsForUsername = (username) => {
```

Comment:
Most of these `lets` should be `const`. Use `const` by default until you need `let`.